docs/design/uptime-reputation-calculation.md

```diff
@@ -0,0 +1,122 @@
+1. The _"score"_ obtained from uptime check failures, without any further modification, doesn't integrate with the current _SN reputation system_.
+
+While one of our goals is to reduce the number of resources used by the satellite we see that the advantages beat the disadvantages considering that the current system got disabled after some nodes got disqualified unfairly and/or unclear.
+
+On the other hand, we think that the number of resources required by this new system shouldn't be too much and we believe that we can have an implementation that scales up.
+
+## Implementation
+
+### DB
+
+A database is required to store information about _SNs_ uptime check failures. This section presents the new required database schema and a brief explanation what it will store.
+
+Only one database table is required and its schema is the following (pseudo code):
+
+```sql
+CREATE TABLE failed_uptime_checks (
+    node_id       BYTEA NOT NULL,
+    when          TIMESTAMP WITH TIMEZONE NOT NULL, -- When the SN failed the an uptime check until it succeeds a future uptime check
+    back_online   TIMESTAMP WITH TIMEZONE, -- When SN has succeeded an uptime check after the last one failed
+    disqualified  TIMESTAMP WITH TIMEZONE  -- When the SN has been disqualified by any disqualification system (uptimes, audits, etc.)
+)
+```
+
+The table will be filled with _SNs_ which have failed at least one uptime check during the ongoing month.
+
+Each row in the table indicates when an _SN_ failed the first uptime check (`when`), while the _SN_ not succeed any future uptime check (`back_online` is `NULL`); when the _SN_ is back online any future uptime check failure will insert a new row to mark a new period of time that the node is offline again.
+
+When the _SN_ is disqualified by any disqualification system and the _SN_ isn't back online, the `disqualified` column of the row will be updated for not receiving more uptime checks because we don't do business with disqualified _SNs_.
+
+### Algorithm
+
+The algorithm is composed of 2 parts:
+
+1. Select which _SNs_ will be checked for uptime, henceforth uptime check _SN_ selection.
+1. Recheck the uptime of the _SNs_ which has failed the initial uptime check (part 1) during an interval period of time, henceforth uptime recheck loop.
+
+#### Uptime Check _SN_ selection
+
+This design doc doesn't change the _SN_ selection process for uptime checks therefore the current implemented mechanism with some minor modifications to classify those nodes which fail the uptime check in order of being recheck by the second part of the algorithm.
+
+Once an _SN_ is selected for an uptime check, the current implemented uptime check mechanism must be modified to achieve the following algorithm:
+
+1. If there is a row in the `failed_uptime_checks` table for the selected _SN_ with `back_online` set to `NULL` do nothing, otherwise follow with the next step.
```

Comment:
Isaac is currently adding a cache of this information, I believe?  Can the document reflect the fact that this information should be cached?  My ideal would be that the majority of uptime / address reporting doesn't even do a SELECT from the database.

Reply:
This has been rephrased in terms of that this operations aren't directly with the DB are through a service interface and such database should only be accessible by the service.

Isaac is adding a cache system for the existing system, not for this one.